File: src/models/GameplaySession.js
const mongoose = require('mongoose');

// Define the schema for a Gameplay Session
const gameplaySessionSchema = new mongoose.Schema({
    sessionId: { type: String, required: true, unique: true },  // Unique session identifier
    userId: { type: String, required: true }, // User ID associated with the session
    startTime: { type: Date, required: true, default: Date.now }, // Session start time
    duration: { type: Number, required: true }, // Duration of the session in seconds
    resolution: { type: String, required: true }, // Resolution of the recording
    frameRate: { type: Number, required: true }, // Frame rate of the recording
    fileSize: { type: Number, required: true }, // Size of the recorded file in MB
    metadata: { type: Object }, // Additional metadata for the session
}, {
    timestamps: true, // Automatically manage createdAt and updatedAt fields
    collection: 'gameplay_sessions', // Collection name in MongoDB
});

// Create the model from the schema
const GameplaySession = mongoose.model('GameplaySession', gameplaySessionSchema);

module.exports = GameplaySession; // Export the model for use in other parts of the application

File: src/models/gameplay_session.py
from pydantic import BaseModel
from typing import List, Optional

class GameplayEvent(BaseModel):
    event_type: str  # Type of event (e.g., 'click', 'keypress')
    timestamp: float  # Time when the event occurred
    details: dict  # Additional details about the event

class GameplaySession(BaseModel):
    session_id: str  # Unique identifier for the gameplay session
    user_id: str  # Identifier for the user who played
    start_time: float  # Session start time
    end_time: float  # Session end time
    events: List[GameplayEvent]  # List of gameplay events
    score: Optional[int]  # Optional score achieved during the session

    def duration(self) -> float:
        """Calculates the duration of the gameplay session."""
        return self.end_time - self.start_time

    def add_event(self, event: GameplayEvent):
        """Adds a new event to the gameplay session."""
        self.events.append(event)
